whylog/constraints/__init__.py

```diff
@@ -2,42 +2,177 @@
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstraintVerificationError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    TYPE = 'undefined'
+    MIN_GROUPS_COUNT = None
+    MAX_GROUPS_COUNT = None
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        Creates constraint object.
+
+        For Teacher and Front use while creating user rule.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
```

Comment:
Yes, you are thinking about right groups.
`group_no` is number of group in regex (or other pattern) , numbering from left, from 1.
`line_id` is id of line in inner numeration between Front and Teacher.
line_id is needed, because in `groups` can be groups from several lines.